### PR TITLE
[KubeRay] Format autoscaling config based on RayCluster CR

### LINKED ISSUE #21655: [Kuberay] Automatically generate Autoscaler YAML config from CRDs
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

At the moment in https://github.com/ray-project/ray/pull/21086, the autoscaler config is manually posted to Kubernetes and shared between Ray clusters (and needs to be manually kept in sync with the Ray cluster CRD). Going forward we should autogenerate the config from the CRD and allow per cluster configs.

I think this is pretty important (P1) since it impacts the user interface, which we should try to stabilize sooner rather than later.

There are two ways to do this:

- In the Kuberay operator, generate the .yaml from the head and worker groups in the RayCluster CRD statically when the cluster is instantiated, create a  the autoscaler yaml as a ConfigMap and mount it into the autoscaler side car. This has the advantage that it is super robust and doesn't increase the number of k8s API requests and keeps kubernetes specific code out of the autoscaler as much as possible.
- Generate the .yaml dynamically in the autoscaler as proposed in https://github.com/ray-project/ray/pull/21086#issuecomment-1013563640. This has the advantage that it supports dynamically updating head and worker groups.

I would vote for the first solution for the added robustness -- dynamically updating groups doesn't seem like an important feature (but per cluster groups and not having to create the yaml files manually definitely is!).

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

Closes https://github.com/ray-project/ray/issues/21655. At the start of each autoscaler iteration, we read the Ray Cluster CR from K8s and use it to extract the autoscaling config.

Some todos outside the scope of this PR:
- There are some rough edges around extracting resource annotations for the autoscaler (especially around custom resource annotations) 
- The code in the docker directory needs to be moved into the python autoscaler code directory
- We should add functionality that automatically formats autoscaler container configuration.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Manually checked it works

### CODE REVIEW

**Comment by DmitriGekhtman** [comment]
cc @Jeffwan

**Comment by DmitriGekhtman** [comment]
@pcmoritz I think the spelling is KubeRay?
https://github.com/ray-project/kuberay#readme

Kuberay is [something else](https://www.youtube.com/watch?v=psQv5wtxNeA) :)

**Comment by DmitriGekhtman** [comment]
@pcmoritz once this is merged, I'll follow up with you merge an updated experimental kuberay-autoscaler image

File: python/ray/autoscaler/_private/kuberay/node_provider.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ def _url(self, path: str) -> str:
             + path
         )
 
-    def _get(self, path: str) -> Dict[str, Any]:
+    def get(self, path: str) -> Dict[str, Any]:
```
Comment (wuisawesome) [inline python/ray/autoscaler/_private/kuberay/node_provider.py:145]: ooc why this change?

File: docker/kuberay-autoscaler/autoscaling_config.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+_RAY_CLUSTER_NAMESPACE = os.getenv("RAY_CLUSTER_NAMESPACE")
+
+# Logical group name for the KubeRay head group.
+# Used as the name of the "head node type" by the autoscaler.
+_HEAD_GROUP_NAME = "head-group"
+
+_GPU_WARNING_LOGGED = False
+
+
+def generate_autoscaling_config() -> dict:
```
Comment (wuisawesome) [inline docker/kuberay-autoscaler/autoscaling_config.py:28]:
```suggestion
def generate_autoscaling_config() -> Dict[str, Any]:
```

File: docker/kuberay-autoscaler/autoscaling_config.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        **legacy_autoscaling_fields,
+    }
+
+    # Make sure the config is readable by the autoscaler.
+    validate_config(autoscaling_config)
+
+    return autoscaling_config
+
+
+def _generate_provider_config() -> dict:
```
Comment (wuisawesome) [inline docker/kuberay-autoscaler/autoscaling_config.py:90]:
```suggestion
def _generate_provider_config() -> Dict[str, Any]:
```

File: docker/kuberay-autoscaler/autoscaling_config.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        "worker_setup_commands": [],
+        "head_start_ray_commands": [],
+        "worker_start_ray_commands": [],
+        "auth": {},
+        "head_node": {},
+        "worker_nodes": {},
+    }
+
+
+def _generate_available_node_types_from_ray_cr_spec(ray_cr_spec: dict) -> dict:
```
Comment (wuisawesome) [inline docker/kuberay-autoscaler/autoscaling_config.py:120]:
```suggestion
def _generate_available_node_types_from_ray_cr_spec(ray_cr_spec: Dict[str, Any]) -> Dict[str, Any]:
```

File: docker/kuberay-autoscaler/autoscaling_config.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+        **{
+            worker_group_spec["groupName"]: _node_type_from_group_spec(
+                worker_group_spec, is_head=False
+            )
+            for worker_group_spec in ray_cr_spec["workerGroupSpecs"]
+        },
+    }
+
+
+def _node_type_from_group_spec(group_spec: dict, is_head: bool) -> dict:
```
Comment (wuisawesome) [inline docker/kuberay-autoscaler/autoscaling_config.py:136]:
```suggestion
def _node_type_from_group_spec(group_spec: Dict[str, any], is_head: bool) -> Dict[str, Any]:
```

File: docker/kuberay-autoscaler/run_autoscaler.py [hunk 6]
```diff
@@ -52,11 +53,12 @@ def setup_logging() -> None:
...
 
-    cluster_name = yaml.safe_load(open(AUTOSCALING_CONFIG_PATH).read())["cluster_name"]
     head_ip = get_node_ip_address()
     Monitor(
         address=f"{head_ip}:6379",
         redis_password=args.redis_password,
-        autoscaling_config=AUTOSCALING_CONFIG_PATH,
+        # The `autoscaling_config` arg can be a dict or a `Callable: () -> dict`.
+        # In this case, it's a callable.
+        autoscaling_config=generate_autoscaling_config,
```
Comment (wuisawesome) [inline docker/kuberay-autoscaler/run_autoscaler.py:62]: Can we add the type hint to `monitor.py` instead of a comment here?

**Review by wuisawesome — APPROVED** [review]

**Comment by wuisawesome** [comment]
I also have one higher level question here. What's the relationship between an autoscaler config and a ray cluster cr? It seems like the relationship should either be

1. Autoscaler config can always be derived from a CR
2. CR can always be derived form an autoscaler config
3. They are 1:1 mappings.

It seems like this PR implies either (1) or (3), just think it's worth explicitly pointing out.

**Comment by pcmoritz** [comment]
@DmitriGekhtman Haha ok, I always found Kuberay to be the prettier spelling but if KubeRay is the official one let's use that :)

File: docker/kuberay-autoscaler/autoscaling_config.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+import json
+
+from ray.autoscaler._private.providers import _get_node_provider
+from ray.autoscaler._private.util import validate_config
+
+logger = logging.getLogger(__name__)
+
+# The name of this Ray cluster. Should coincide with the `metadata.name` field of the
+# Ray CR.
+_RAY_CLUSTER_NAME = os.getenv("RAY_CLUSTER_NAME")
```
Comment (pcmoritz) [inline docker/kuberay-autoscaler/autoscaling_config.py:16]: Can we shift this (and the _RAY_CLUSTER_NAMESPACE below) to a command line argument instead of an environment variable? That would be less error prone, since it can't be overwritten by accident and the control flow of how things get passed through is clearer.

File: python/ray/autoscaler/_private/kuberay/node_provider.py (hunk 1 above)
Comment (pcmoritz) [reply python/ray/autoscaler/_private/kuberay/node_provider.py:145]: +1, I think keeping it private makes more sense :)

Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/kuberay/node_provider.py:145]: It's currently used in `autoscaling_config.py` to fetch the RayCluster CR, therefore is not private at the moment.
I'll just use the Python K8s client there instead -- we don't need to see the node provider in that file.

File: docker/kuberay-autoscaler/autoscaling_config.py (hunk 2 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/autoscaling_config.py:28]: Oh yeah, I always forget that dicts can have keys other than strings.

File: docker/kuberay-autoscaler/run_autoscaler.py (hunk 6 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/run_autoscaler.py:62]: I will keep the comment and also add the type hint.

File: docker/kuberay-autoscaler/autoscaling_config.py (hunk 7 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/autoscaling_config.py:16]: Good idea, will do.

**Comment by DmitriGekhtman** [comment]
> I also have one higher level question here. What's the relationship between an autoscaler config and a ray cluster cr? It seems like the relationship should either be
> 
> 1. Autoscaler config can always be derived from a CR
> 2. CR can always be derived form an autoscaler config
> 3. They are 1:1 mappings.
> 
> It seems like this PR implies either (1) or (3), just think it's worth explicitly pointing out.

It is (1). The autoscaler knows nothing of pod configuration, which is the bulk of the CR; `node_config: {}` is vestigial. 
I think the relationship can be made clearer by splitting the autoscaling config fetch as 
`def fetch_ray_cr(cluster_name, cluster_namespace)`
`def derive_autoscaling_config_from_ray_cr(ray_cr)`

File: docker/kuberay-autoscaler/autoscaling_config.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+
+    # Get the KubeRay node provider by instantiating it (on first call) or retrieving it
+    # from cache (subsequent calls).
+    # The autoscaler accesses the single node provider instance in the same way.
+    provider = _get_node_provider(
+        provider_config=provider_config, cluster_name=_RAY_CLUSTER_NAME, use_cache=True
+    )
+
+    # Fetch the Ray CR from K8s.
+    ray_cr = provider.get("rayclusters/{}".format(_RAY_CLUSTER_NAME))
```
Comment (Jeffwan) [inline docker/kuberay-autoscaler/autoscaling_config.py:54]: how does the provider detect the namespace?  

provider.get("rayclusters/{}/{}".format(_RAY_CLUSTER_NAMESPACE, _RAY_CLUSTER_NAME))

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml [hunk 9]
```diff
@@ -111,6 +115,16 @@ spec:
...
+          env:
+          - name: RAY_CLUSTER_NAMESPACE
+            valueFrom:
+              fieldRef:
+                fieldPath: metadata.namespace
+          - name: RAY_CLUSTER_NAME
+            # This value must match the metadata.name of the RayCluster CR.
+            # The autoscaler uses this env variable to determine which Ray CR to interact with.
+            # TODO: Match with CR name automatically via operator, Helm, and/or Kustomize.
+            value: raycluster-complete
```
Comment (Jeffwan) [inline python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:127]: why not to use fieldRef as namespace?

File: docker/kuberay-autoscaler/autoscaling_config.py [hunk 10]
```diff
@@ -0,0 +1,282 @@
+from contextlib import suppress
```
Comment (Jeffwan) [inline docker/kuberay-autoscaler/autoscaling_config.py:1]: with this configuration, we can remove configmap. https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/kuberay/kuberay-autoscaler.yaml#L1

**Comment by Jeffwan** [comment]
I will help on the orchestration part in kuberay side once this is finalized the merged

File: docker/kuberay-autoscaler/autoscaling_config.py (hunk 8 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/autoscaling_config.py:54]: I'm going to replace this part with a call to the Python K8s client.
Should be easier to read.

File: docker/kuberay-autoscaler/autoscaling_config.py (hunk 10 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/autoscaling_config.py:1]: Right, thanks for reminding me to `rm` that.

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml (hunk 9 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:127]: `fieldRef` supports certain fields of a pod
I don't know of a good way to reference the name of the CR.

The fieldRef used for the namespace is a reference to the namespace of the Ray head pod, which happens to be the same as the name of the RayCluster CR.

**Comment by Jeffwan** [comment]
Originally, I was thinking to ask operator to do the job to covert CRD to configmap and mount to autoscaler container.  Current way still works and simplify orchestration and management (e.g. no more configmaps). There's one limitation that current autoscaler won't watch crd and update the configuration if there's a change to the cluster.  I think this is fine at this moment.

**Comment by DmitriGekhtman** [comment]
> Originally, I was thinking to ask operator to do the job to covert CRD to configmap and mount to autoscaler container. Current way works as well and it simplifies orchestration and management (e.g. no more configmaps). There's one limitation that current autoscaler won't watch crd and update the configuration if there's a change to the cluster. I think this is fine at this moment.

At the moment, there isn't an improvement to be had by having the autoscaler put a watch on the CR. With the current design, periodic reads are sufficient.

The autoscaler polls all relevant state and configuration once per set interval (default 5 seconds). It then uses that fixed info to execute one iteration of scaling.

As a result, there's a possibility of brief edge effects when the CR is update during an autoscaler update. Things would get to a consistent state at the time of the next autoscaler update.

File: docker/kuberay-autoscaler/autoscaling_config.py (hunk 8 above)
Comment (pcmoritz) [reply docker/kuberay-autoscaler/autoscaling_config.py:54]: Here is what I would do about this: Define a class
```
class AutoscalingConfigGenerator:
    def __init__(self):
        self.headers, self.verify = ray.autoscaler._private.kuberay.node_provider.load_k8s_secrets() # factor this method out

    def __call__(self, args):
        url = ray.autoscaler._private.kuberay.node_provider.url_from_resource("rayclusters/...") # factor out
        requests.get(url, headers=self.headers, verify=self.verify)
        # generate autoscaling config here
```
and then pass an instance of this class (which is a callable) to autoscaling_config of Monitor. This way, the secrets only have to be read once (vs. having to instantiate lots of providers or k8s clients, which is even worse than creating lots of providers) and there is maximum code sharing with the node provider/everything is consistent.

Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/autoscaling_config.py:54]: Was planning to just instantiate a single extra global k8s client for this purpose, but sure.

File: docker/kuberay-autoscaler/autoscaling_config.py (hunk 7 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/autoscaling_config.py]: Kept the envs and passed them as command line arguments to avoid including "default" namespace in the command line args.
The command line arguments can't be overwritten at runtime.

File: docker/kuberay-autoscaler/autoscaling_config.py (hunk 8 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/autoscaling_config.py]: Created the class, named it `AutoscalingConfigProducer` to avoid suggesting that it's a Python generator. A few lines of code duplicated, but it's not so bad.

**Comment by DmitriGekhtman** [comment]
Should be close to final.

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml [hunk 11]
```diff
@@ -22,15 +22,22 @@ spec:
     # rayNodeType: head # Not needed since it is under the headgroup
     # the following params are used to complete the ray start: ray start --head --block --redis-port=6379 ...
     rayStartParams:
+      # Flag "no-monitor" must be set when running the autoscaler in
+      # a sidecar container.
+      no-monitor: "true"
```
Comment (chenk008) [inline python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:27]: Can we run autoscaler in a sidecar container currently?  Maybe we need to update the Raycluster CRD too.

Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:27]: This example CR sets up the autoscaler to run in a sidecar.

In future work, we should update the operator and custom resource to automatically configure the autoscaler sidecar
(or conceivably an independent autoscaler deployment, if it's desired).

File: python/ray/autoscaler/_private/kuberay/node_provider.py [hunk 12]
```diff
@@ -90,6 +90,49 @@ def make_node_tags(labels: Dict[str, str], status_tag: str) -> Dict[str, str]:
...
+
+    headers = {
+        "Authorization": "Bearer " + token,
+    }
+    verify = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+
+    return headers, verify
+
+
+def url_from_resource(namespace: str, path: str) -> str:
```
[the remark this answers is not in the document]
Comment (DmitriGekhtman) [reply python/ray/autoscaler/_private/kuberay/node_provider.py:112]: The Python client's `patch_namespaced_custom_resource` method has bugs, so we decided to use the REST API directly for that. For consistency, we've kept to the REST API throughout.

We're open to changing this down the line. Context: https://github.com/ray-project/ray/pull/21086#discussion_r782726190

I think CRD version management is a concern whether or not we use the Python client.
That's something we should definitely keep in mind, but I think it's out of scope for this PR.

We're still working in the context of a prototype, meant to be deployed with fixed versions for everything, as described here:
https://docs.ray.io/en/master/cluster/kuberay.html

File: docker/kuberay-autoscaler/Dockerfile [hunk 13]
```diff
@@ -4,5 +4,6 @@ RUN $HOME/anaconda3/bin/pip uninstall -y ray
 COPY $WHEEL_PATH .
 RUN $HOME/anaconda3/bin/pip --no-cache-dir install "$WHEEL_PATH"[all]
 
+COPY run_autoscaler_with_retries.py /home/ray/run_autoscaler_with_retries.py
```
Comment (pcmoritz) [inline docker/kuberay-autoscaler/Dockerfile:7]: This line looks like a duplicate with the line two below

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml [hunk 14]
```diff
@@ -97,8 +104,21 @@ spec:
...
+          env:
+          - name: RAY_CLUSTER_NAMESPACE
+            valueFrom:
+              fieldRef:
+                fieldPath: metadata.namespace
+          - name: RAY_CLUSTER_NAME
+            # This value must match the metadata.name of the RayCluster CR.
+            # The autoscaler uses this env variable to determine which Ray CR to interact with.
+            # TODO: Match with CR name automatically via operator, Helm, and/or Kustomize.
+            value: raycluster-complete
```
Comment (pcmoritz) [inline python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:116]: Why not
```
valueFrom:
  fieldRef:
    fieldPath: metadata.name
```
here and get rid of the TODO?

Comment (pcmoritz) [reply python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:116]: Ah I see, this is the pod template not the RayCluster spec. Probably the best solution here is to let kuberay define this environment variable. @DmitriGekhtman and @Jeffwan can you make sure this happens? I don't want to end up with a worse solution here just because the code is living in two different repositories.

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml [hunk 15]
```diff
@@ -97,8 +104,21 @@ spec:
...
+          command: ["/bin/sh", "-c"]
```
Comment (pcmoritz) [inline python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:117]: Would it be possible to do
```
command: /home/ray/anaconda3/bin/python
args:
- "/home/ray/run_autoscaler_with_retries.py"
- "--cluster-name"
- "$(RAY_CLUSTER_NAME)"
- "--cluster-namespace"
- "$(RAY_CLUSTER_NAMESPACE)"
- "--redis-password"
- "5241590000000000"
```
and avoid the additional shell? See also https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/#use-environment-variables-to-define-arguments

File: python/ray/autoscaler/kuberay/kuberay-autoscaler.yaml [hunk 16]
```diff
@@ -1,70 +0,0 @@
...
-    setup_commands: []
-    head_setup_commands: []
-    worker_setup_commands: []
-    head_start_ray_commands: []
-    worker_start_ray_commands: []
-    auth: {}
-    head_node: {}
-    worker_nodes: {}
----
-apiVersion: v1
```
Comment (pcmoritz) [inline python/ray/autoscaler/kuberay/kuberay-autoscaler.yaml:40]: This is important: The service account, role and role binding definitions are still needed.

Related, what is the plan for doing an end-to-end test of the autoscaler in the Ray CI? Do we already have a PR for that? It would have caught this problem.

File: python/ray/autoscaler/_private/monitor.py [hunk 17]
```diff
@@ -139,11 +140,11 @@ class Monitor:
 
     def __init__(
         self,
-        address,
-        autoscaling_config,
-        redis_password=None,
-        prefix_cluster_info=False,
-        monitor_ip=None,
+        address: str,
+        autoscaling_config: Union[str, Callable[[], Dict[str, Any]]],
```
Comment (pcmoritz) [inline python/ray/autoscaler/_private/monitor.py:144]: Nice!

File: docker/kuberay-autoscaler/autoscaling_config.py [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+
+
+def _get_custom_resources(
+    ray_start_params: Dict[str, Any], group_name: str
+) -> Dict[str, int]:
+    if "resources" not in ray_start_params:
+        return {}
+    resources_string = ray_start_params["resources"]
+    try:
+        resources_json = resources_string[1:-1].replace("\\", "")
```
Comment (pcmoritz) [inline docker/kuberay-autoscaler/autoscaling_config.py:286]: Can you document why the indexing and replacing is needed here? I assume it is related to the unfortunate resource specification below :)

File: docker/kuberay-autoscaler/run_autoscaler.py [hunk 19]
```diff
@@ -52,11 +68,23 @@ def setup_logging() -> None:
...
+
+    autoscaling_config_producer = AutoscalingConfigProducer(
+        args.cluster_name, args.cluster_namespace
+    )
+
+    # K8s might not consider the Ray head pod to be `Running` at the moment that
+    # this entrypoint code runs.
+    # That can lead to a annoying artifact at the start of the autoscaler logs.
+    # Wait a bit to avoid that artifact.
+    time.sleep(5)
```
Comment (pcmoritz) [inline docker/kuberay-autoscaler/run_autoscaler.py:81]: What would we need to do to get rid of this sleep and still avoid the artifacts in the autoscaler logs? I think we should try to address that as part of this PR.

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml (hunk 14 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:116]: Yep, definitely on the todo to automate formatting of the autoscaler container spec!

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml (hunk 15 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:117]: oh great point, let's do that

File: python/ray/autoscaler/kuberay/kuberay-autoscaler.yaml (hunk 16 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/kuberay-autoscaler.yaml:40]: Thanks for catching that! Only meant to remove the configMap.

Plan: I'll write tests relatively soon that can be run against any K8s cluster.
@simon-mo is working on the CI infra to automated with minikube with a target of ~3/15.

File: docker/kuberay-autoscaler/autoscaling_config.py (hunk 18 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/autoscaling_config.py:286]: Indeed, most unfortunate formatting for the unfortunate resource specification.

File: docker/kuberay-autoscaler/run_autoscaler.py (hunk 19 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/run_autoscaler.py:81]: We can have the node provider wait for the head node to be registered as running during the node provider's initialization.

same wait, more tightly organized

File: docker/kuberay-autoscaler/Dockerfile (hunk 13 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/Dockerfile:7]: Indeed, I had meant to copy "autoscaling_config.py".

File: docker/kuberay-autoscaler/run_autoscaler.py (hunk 19 above)
Comment (pcmoritz) [reply docker/kuberay-autoscaler/run_autoscaler.py:81]: Yes, let's do that, that sounds like the right solution :)

**Comment by DmitriGekhtman** [comment]
> Thanks for doing all this work! If you address the remaining comments this looks good to be merged as a first cut.
> 
> As a next step, we should try to improve the interfaces between KubeRay and Ray (e.g. around ray startup parameters and resource specifications). Let's write a design doc about it and discuss it in the KubeRay meetings. Maybe @DmitriGekhtman, @Jeffwan and I can work together on the design doc?

sgtm

File: python/ray/autoscaler/kuberay/ray-cluster.complete.yaml (hunk 14 above)
Comment (DmitriGekhtman) [reply python/ray/autoscaler/kuberay/ray-cluster.complete.yaml:116]: Resolving thread for my own tracking purposes.

File: docker/kuberay-autoscaler/run_autoscaler.py (hunk 19 above)
Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/run_autoscaler.py:82]: After a closer examination, it seems to relate to the Ray internals.
I've updated the explanation.

We should probably modify the autoscaler code to work around it, but not right now.

Comment (DmitriGekhtman) [reply docker/kuberay-autoscaler/run_autoscaler.py:82]: Yeah, ok, I see, the solution is a minor refactor of `StandardAutoscaler.summary` but best left for later.

**Comment by DmitriGekhtman** [comment]
Test failures are unreleated.
